src/allmydata/storage/http_server.py

```diff
@@ -312,21 +413,15 @@ def read_range(request, read_data: Callable[[int, int], bytes]) -> None:
         raise _HTTPError(http.REQUESTED_RANGE_NOT_SATISFIABLE)
 
     offset, end = range_header.ranges[0]
-
-    # TODO limit memory usage
-    # https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3872
-    data = read_data(offset, end - offset)
-
     request.setResponseCode(http.PARTIAL_CONTENT)
-    if len(data):
-        # For empty bodies the content-range header makes no sense since
-        # the end of the range is inclusive.
-        request.setHeader(
-            "content-range",
-            ContentRange("bytes", offset, offset + len(data)).to_header(),
-        )
-    request.write(data)
-    request.finish()
+    d = Deferred()
+    request.registerProducer(
+        _ReadRangeProducer(
+            request, read_data_with_error_handling, d, offset, end - offset
+        ),
+        False,
+    )
+    return d
```

Comment:
The return annotation needs to be updated.

Reply:
Yep. And... `tox -e typechecks` didn't catch it.